### PR TITLE
Makes plasma cabbages absorb plasma from the air

### PR DESCRIPTION
**Pull request by joaomarcosh**
## What this does
Makes plasma cabbages (both the plant and the products) absorb plasma from the air.
If the product absorbs plasma it will increase the amount of plasma reagent inside of it.
if the plant absorbs plasma it increases potency.

As it currently stands, 10 plasma cabbages can clear a room filled with an entire plasma canister in around 6 minutes. Pic below is the setup i used to measure the time.

## Why it's good
Adds a mechanic to an item that no one uses and makes botany ever so slightly more useful in case of plasma leaks.

## Changelog
:cl:
 * rscadd: Plasma cabbages can now absorb plasma from the air
 * rscadd: Add a xenophysiology trait that absorbs gasses from the air and turn it into reagents

### CODE REVIEW

File: code/modules/hydroponics/hydro_tray_process.dm [hunk 1]
```diff
@@ -235,6 +235,12 @@
 				continue
 			environment.adjust_gas(gas, -(seed.consume_gasses[gas]), FALSE)
 		environment.update_values()
+		
+	// Special case for plasma cabbage plants to absorb plasma from the air
+	// doesn't affect seed/product potency because i can't figure out how to use check_for_divergence()
+	if(istype(seed,/datum/seed/plasmacabbage) && environment[GAS_PLASMA] > 0)
+		environment.adjust_gas(GAS_PLASMA, min(-2,environment[GAS_PLASMA]), FALSE)
```
Comment (pedr69) [inline code/modules/hydroponics/hydro_tray_process.dm:242]: why the `min()`?
it'll always be -2, won't it?

Comment (joaomarcosh) [reply code/modules/hydroponics/hydro_tray_process.dm]: its supposed to be min(-2, -environment[GAS_PLASMA]) so the plasma doesn't go negative (it happened during my tests)

**Comment by Eneocho** [comment]
I suggest you make this a trait under xenophysiology instead of it being snowflaked to plasma cabbages, such that you could splice the consumption to other plants.
Of note is that the gas consumption trait already exists, and can consume plasma. Issue is that if the plant does NOT consume said gas, it starts to slowly die. This does NOT affect the crop.
>makes botany ever so slightly more useful in case of plasma leaks.

You can use the aforementioned gas consumption trait to deal with plasma leaks. It is a chore to find with the correct gas consumption (such that it eats plasma and not, say, oxygen), though, so this is a welcome change.
This could instead be "Gas Capture" or similar. It takes a specific gas from the environment and slowly turns it into reagent inside the fruit (Oxygen with O2, Nitrogen with N2, Plasma with Plasma, Oxygen and Carbon in a 2:1 rate with CO2, Oxygen and Nitrogen in a 2:1 rate with NO2), like a reverse Vapor Sac.

File: code/modules/hydroponics/seeds.dm [hunk 2]
```diff
@@ -1162,7 +1162,7 @@
 	lifespan = 50
 	maturation = 3
 	production = 5
-	yield = 4
+	yield = 1
```
Comment (Eneocho) [inline code/modules/hydroponics/seeds.dm]: This might be an accident. You reduced the yield of regular cabbage.
Did you intend to change the yield of plasma cabbage? If so, put it in the changelog under tweak.
As in:
* tweak: Plasma cabbages now make only 1 cabbage per harvest.

I'd keep the yield reduction to its own PR, though, for atomicity.

**Comment by MercuriusGH** [comment]
Another win for botany warriors

File: code/modules/reagents/reagent_containers/food/snacks/grown.dm [hunk 3]
```diff
@@ -407,9 +407,40 @@ var/list/strange_seed_product_blacklist = subtypesof(/obj/item/weapon/reagent_co
...
+	if(!environment && istype(T))
+		environment = T.return_air()
+	else
+		environment = space_gas
+		
+	if(environment[GAS_PLASMA] > 0 && current_capacity < max_capacity)
+		var/amount_consumed = min(min(environment[GAS_PLASMA],consume_amount),min(consume_amount,max_capacity-current_capacity))
+		current_capacity += amount_consumed
+		potency += 0.1
+		reagents.reagent_list[2].volume += 0.1
```
Comment (pedr69) [inline code/modules/reagents/reagent_containers/food/snacks/grown.dm:438]: recommend you `add_reagent()` the plasma instead, the current way would let you dupe chems after purging the plasma fron the cabbage
i'm all for letting botany duplicate exotic chems if they can get a sample but it should be done in a more intuitive way than waterpotass bullshittery

Comment (joaomarcosh) [reply code/modules/reagents/reagent_containers/food/snacks/grown.dm]: I actually tried doing it that way but no matter what i did it never increased it. I will do another pass and see if i can make that work and turn this into a trait.

**Comment by Hinaichigo** [comment]
If it gets full of plasma or other reagents, what happens?

**Comment by Eneocho** [comment]
Make it so you can find all versions of the gas absorbing trait on exotic seeds, as in, co2 absorbing, o2 absorbing, etc.
Also, does the new trait appear on the plant scanner? Does it have a shorthand on the seed extractor menu? If not, it should.

**Comment by joaomarcosh** [comment]
> If it gets full of plasma or other reagents, what happens?

It stops absorbing.


> Make it so you can find all versions of the gas absorbing trait on exotic seeds, as in, co2 absorbing, o2 absorbing, etc. Also, does the new trait appear on the plant scanner? Does it have a shorthand on the seed extractor menu? If not, it should.

I made it so if the exotic seed generates with a consume_gas it will have 10% chance to also roll the gas_absorb trait.
I also added the trait to the plant analyzer and seed extractor.

**Comment by joaomarcosh** [comment]
Unless the reviews have anything else to say, I believe that was the last commit.
Maybe we could change some of the numbers but I rather live test it first.

**Comment by Eneocho** [comment]
>I made it so if the exotic seed generates with a consume_gas it will have 10% chance to also roll the gas_absorb trait.

I'd make it independent, cause you cannot find the consume gas trait on its own from the seeds as it is now. Do you intend for the two traits to be linked or be mostly independent besides both dealing in gas shenanigans? If you want them to be interdependent, you could make it so gas absorb is a child mutation of gas consume (like how juicy evolves into slippery, but are the same trait behind the curtain, slippery is just Juicy = 2).
As it is now though, your only sole source would be the plasma absorb from the plasma cabbage (which you can't use for other gases), or mutation (which is not the most reliable, nor safe, method). The odds would be about 0.5% to find per seed, meaning that a sizeable 20 crates purchase would only get you around a 9.3% chance to find an absorbing plant. That's rarer than finding the noreact trait (or any trait, for that matter) on an exotic seed.

**Comment by joaomarcosh** [comment]
> I'd make it independent, cause you cannot find the consume gas trait on its own from the seeds as it is now. 

I didn't make it independent because the gas_absorb trait uses the same list of gases as consume_gasses, so if i make it independent you may have a gas_absorb plant that doesnt actually absorb anything.

>As it is now though, your only sole source would be the plasma absorb from the plasma cabbage (which you can't use for other gases)

I believe you can splice the plasma cabbage xenophysiology trait into another seed that has consume gasses and that would make it absorb both the plasma and the seeds gasses. For example, the chicken of the stars consumes nitrogen so if you splice plasma cabbages onto it you can make it consume and absorb both plasma and nitrogen.
Although if you want something that does not absorb plasma, it will be just exotic seed lottery.

File: code/modules/reagents/reagent_containers/food/snacks/grown.dm [hunk 4]
```diff
@@ -260,6 +262,33 @@ var/list/special_fruits = list()
 		spawn()
 			spark(M) //Two set of sparks, one before the teleport and one after. //Sure then ?
 	return 1
+	
+/obj/item/weapon/reagent_containers/food/snacks/grown/process()
+	var/turf/T = get_turf(src)
+	var/datum/gas_mixture/environment
+	if(!environment && istype(T))
```
Comment (Kurfursten) [inline code/modules/reagents/reagent_containers/food/snacks/grown.dm]: !environment always evaluates true

File: code/modules/reagents/reagent_containers/food/snacks/grown.dm [hunk 5]
```diff
@@ -26,12 +26,14 @@ var/list/special_fruits = list()
...
 
-/obj/item/weapon/reagent_containers/food/snacks/grown/New(atom/loc, custom_plantname, mob/harvester)
+/obj/item/weapon/reagent_containers/food/snacks/grown/New(atom/loc, custom_plantname, mob/harvester, add_process=0)
 	..()
 	if(custom_plantname)
 		plantname = custom_plantname
 	if(ticker)
 		initialize(harvester)
+	if(add_process)
+		processing_objects.Add(src)
```
Comment (Kurfursten) [inline code/modules/reagents/reagent_containers/food/snacks/grown.dm]: given that botany is best known for producing loads of plants, particularly with some tweaks, this has potential to be performance impacting since you can make fucktons of processing objects very quickly; it's not something I would bar a merge over but we need to keep an eye on it

File: code/modules/hydroponics/hydro_tray_process.dm [hunk 6]
```diff
@@ -227,12 +227,17 @@
...
 		missing_gas = 0
 		for(var/gas in seed.consume_gasses)
 			if(environment[gas] < seed.consume_gasses[gas])
-				missing_gas++
+				if(!seed.gas_absorb)
+					missing_gas++
 				continue
+			if (seed.gas_absorb)
+				seed = seed.diverge(1)
+				seed.potency += 0.2
```
Comment (Kurfursten) [inline code/modules/hydroponics/hydro_tray_process.dm:240]: lack of a cap or softcap here is possibly bad given other potency methods do incorporate those
although it would be a fair amount of work to farm cabbages in this way, not all gas consuming plants are so dangerous - for example, vox chicken of the stars consumes nitrogen and it's going to be naturally self-buffing at all times

File: code/modules/reagents/reagent_containers/food/snacks/grown.dm (hunk 5 above)
Comment (joaomarcosh) [reply code/modules/reagents/reagent_containers/food/snacks/grown.dm]: Is there another way to do this besides using the processing list or is it just an issue with the nature of the feature?

Comment (Kurfursten) [reply code/modules/reagents/reagent_containers/food/snacks/grown.dm]: It's just in its nature, unless you want to pulse less often, but I don't think you do. This is just an inherent thing to consider when you make a call every 2 seconds on something that can be made in bulk

Comment (Eneocho) [reply code/modules/reagents/reagent_containers/food/snacks/grown.dm]: A high 'tism splice can shit 13-15 plants every 20 seconds or so (even faster if you take the time to mutate, but no-one does). Plant 10 of this splice, get 390-450 crops each minute. You can reach, and far surpass those numbers when doing high intensity botany autismo, such as extreme crop farming with selfharvest (or minerals farming), or high yield luck farming via clovers. Keep in mind that that would be the tismo splice, not the grass + chanterelle splice, which is faster, but way less efficient, and the more used splice setup.
Test how laggy it'd get with 400 process crops. Then 800, etc. Then we can see what would be a sane hardcap for the process checks.

Comment (Eneocho) [reply code/modules/reagents/reagent_containers/food/snacks/grown.dm]: I suggest, for testing, that you VV a plasma cabbage seed, go down to the seed var, access it, set it to potency 200, yield 50 and production 1, then plant. That way you can easily spawn the crops in 50 crop chunks in the "natural" (not adminspawned) way they'd spawn in game.

File: code/modules/hydroponics/hydro_tray_process.dm (hunk 6 above)
Comment (Eneocho) [reply code/modules/hydroponics/hydro_tray_process.dm:240]: Funnily enough, this would be the only legitimate way to get 200 potency instead of the scuffed 198 we have currently, cause my ghetto fix to just bump the potency to 200 was rejected, and I don't understand DM wizardry enough to tweak the formula correctly.
